architecture/2024-05-radius-on-dapr.md

```diff
@@ -0,0 +1,324 @@
+| State store        | A key-value-store building block. See [docs](https://docs.dapr.io/reference/components-reference/supported-state-stores/) for supported state stores.                                                                                                                                                                                                                 |
+| Workflows          | A [workflow-as-code](https://docs.dapr.io/developing-applications/building-blocks/workflow/workflow-overview/) programming model for reliable distributed applications.                                                                                                                                                                                               |
+
+
+## Objectives
+
+> **Issue Reference:** <!-- (If appropriate) Reference an existing issue that describes the feature or bug. -->
+
+### Goals
+
+- Adopt Dapr for our data storage and messaging needs. 
+  - Keep `rad init` and `rad install` simple by including Dapr and a default choice of components as part of our Helm chart.
+  - Keep the Helm chart flexible by documenting how to BYO components.
+  - Provide production-guidance for Radius based on Dapr's existing guidance.
+- Remove our existing messaging and storage implementations in favor of Dapr.
+
+### Non goals
+
+- (out of scope) User-facing behavior changes other than installation/dependencies.
+- (out of scope/follow-up) Adoption of Dapr workflows.
+  - While this is highly valuable for us, this 
+
+### User scenarios (optional)
+
+#### Wendy tries Radius out
+
+Wendy wants to learn more about Radius, so she fires up a Kubernetes cluster and tries some of our tutorials. She runs `rad init` to install Radius, which includes Dapr and some components that Radius can use for its storage. This is easy for her because she didn't need to learn about Dapr to try Radius out.
+
+*This scenario is here to point out that the simplicity of the current user experience is maintained.*
+
+#### Wendy and Jake put Radius in production
+
+Wendy's been successful experimenting with Radius, and she wants to take it to production. She works with Jake to set up some cloud datastores for Radius to use, to ensure that data is persisted. Jake creates a DynamoDB table and the necessary SNS/SQS resources, following the Dapr documentation. Jake reads the Radius documentation, installs Dapr, and configures the `Component` objects in Kubernetes that are required. Lastly, Jake runs `helm install radius radius -n radius-system` to install Radius without its default component configuration.
```

Comment:
> Jake creates a DynamoDB table and the necessary SNS/SQS resources, following the Dapr documentation

Shouldn't this be independent of Dapr, why do they need to follow Dapr docs?

> Jake reads the Radius documentation, installs Dapr

I understood dapr installation would be done by default

Reply:
> Shouldn't this be independent of Dapr, why do they need to follow Dapr docs?

Some of the Dapr components require/recommend specific settings in cloud resources. The point I want to make is that we're not responsible for this in Radius, it's part of an existing project. Any ideas to make this more clear?

> I understood dapr installation would be done by default

There are two scenarios:

- `rad init`/`rad install kubernetes` will include Dapr & Contour (not recommended for production)
- `helm install radius ...` will NOT include Dapr & Contour (recommended for production)

This is consistent with our existing behavior for a dependency (Contour)